### PR TITLE
Add _reduce_scatter_base to ProcessGroupWrapper.

### LINKED ISSUE #66329: Add _reduce_scatter_base and _allgather_base to processGroupWrapper
_reduce_scatter_base and _allgather_base are used in FSDP code base, it is good to add those APIs into processGroupWrapper as well, so that these two collectives can be debugged in DETAIL mode

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by kumpera**
Fixes #66329

### CODE REVIEW

**Review by fduwjj — APPROVED** [review]
lgtm, just a high-level n00b question. What's the purpose of this wrapper? Also you want to let linter happy.

File: test/distributed/test_pg_wrapper.py [hunk 1]
```diff
@@ -265,27 +265,72 @@ def test_collective_hang(self):
...
+            with self.assertRaisesRegex(RuntimeError, ".*") as cm:
+                output = torch.zeros(4 + self.rank, device=device)
+                input = torch.ones(4 * self.world_size, device=device)
+                if self.rank == 0:
+                    wrapper_pg._allgather_base(output, input).wait()
+                else:
+                    wrapper_pg._reduce_scatter_base(output, input).wait()
+            self._validate_error(
+                exception=cm.exception,
+                op_type="ALLGATHER_BASE" if self.rank == 0 else"REDUCE_SCATTER_BASE",
```
Comment (rohan-varma) [inline test/distributed/test_pg_wrapper.py:305]: strange, doesn't look like there is a space after `else` before `REDUCE_SCATTER_BASE`, wonder why python is not complaining

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.hpp [hunk 2]
```diff
@@ -111,6 +111,11 @@ class TORCH_API ProcessGroupWrapper : public ProcessGroup {
   c10::intrusive_ptr<ProcessGroup::Work> barrier(
       const BarrierOptions& opts = BarrierOptions()) override;
 
+    c10::intrusive_ptr<ProcessGroup::Work> _reduce_scatter_base(
+      at::Tensor& outputBuffer,
+      at::Tensor& inputBuffe,
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupWrapper.hpp:116]: `inputBuffer`? (wonder why compiler doesn't complain as it mismatches the cpp argument)

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

@fduwjj The purpose of this wrapper is to work with TORCH_DISTRIBUTED_DEBUG in order to detect collective mismatches.

File: test/distributed/test_pg_wrapper.py (hunk 1 above)
Comment (kumpera) [reply test/distributed/test_pg_wrapper.py:305]: I believe we don't lint the test suite.

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.hpp (hunk 2 above)
Comment (kumpera) [reply torch/csrc/distributed/c10d/ProcessGroupWrapper.hpp:116]: Fixed. arg names have no meaning in forward declarations.

**Comment by kumpera** [comment]
I manually verified that the new failures are unrelated.

**Pull request closed without merging**